src/conn/conn_dhandle.c

```diff
@@ -180,45 +181,87 @@ __wt_conn_btree_sync_and_close(WT_SESSION_IMPL *session, bool final, bool force)
 	 */
 	__wt_spin_lock(session, &dhandle->close_lock);
 
-	/*
-	 * The close can fail if an update cannot be written, return the EBUSY
-	 * error to our caller for eventual retry.
-	 *
-	 * If we are forcing the close, just mark the handle dead and the tree
-	 * will be discarded later.  Don't do this for memory-mapped trees: we
-	 * have to close the file handle to allow the file to be removed, but
-	 * memory mapped trees contain pointers into memory that will become
-	 * invalid if the mapping is closed.
-	 */
+	/* Reset the tree's eviction priority (if any). */
+	__wt_evict_priority_clear(session);
+
+	discard = marked_dead = false;
 	if (!F_ISSET(btree,
 	    WT_BTREE_SALVAGE | WT_BTREE_UPGRADE | WT_BTREE_VERIFY)) {
-		if (force && (bm == NULL || !bm->is_mapped(bm, session))) {
-			F_SET(session->dhandle, WT_DHANDLE_DEAD);
+		/*
+		 * If the object is already marked dead, we're just here to
+		 * discard it.
+		 */
+		if (F_ISSET(dhandle, WT_DHANDLE_DEAD))
+			discard = true;
+
+		/*
+		 * If a standard object and it's OK to mark the handle dead
+		 * (letting the tree be discarded later), do so. Don't mark
+		 * memory-mapped trees dead (we close the file handle to allow
+		 * the file to be removed, but mapped trees contain pointers
+		 * into memory that become invalid if the mapping is closed).
+		 */
+		if (!discard && force && !final &&
+		    (bm == NULL || !bm->is_mapped(bm, session)))
 			marked_dead = true;
 
-			/* Reset the tree's eviction priority (if any). */
-			__wt_evict_priority_clear(session);
-		}
-		if (!marked_dead || final) {
-			if ((ret = __wt_checkpoint_close(
-			    session, final)) == EBUSY)
-				WT_ERR(ret);
-			else
-				WT_TRET(ret);
+		/*
+		 * Get rid of any durable objects we couldn't mark dead. Durable
```

Comment:
We're actually not discarding the handle here, we're discarding the underlying object.